### ISSUE TITLE
Property getter not getting inlined

### ISSUE THREAD

**Issue by Barsonax** [dotnet/coreclr#16119]
I ran into a performance issue when I used a property for the HCost + GCost field [here](https://github.com/Barsonax/Pathfindax/blob/5af5686de82a6307bb54ffdfd3d4343c32e38d92/Source/Code/Pathfindax/Nodes/AstarNode.cs#L43). One would expect these to be inlined as this makes a big difference in performance in my case and the body is really small.

Without the property the perf numbers are this (benchmarks can be run using [this](https://github.com/Barsonax/Pathfindax/blob/5af5686de82a6307bb54ffdfd3d4343c32e38d92/Source/Code/Pathfindax.Benchmarks/Program.cs)):
![image](https://user-images.githubusercontent.com/19387223/35611662-0eb0e9b0-0666-11e8-8be5-424e82bad5a3.png)

But if I replace the HCost + GCost code with a FCost property that returns HCost + GCost I get this:
![image](https://user-images.githubusercontent.com/19387223/35611471-66ae47c6-0665-11e8-9fb0-8624f4e2cde3.png)

Then I decided to look at the generated assembly and noticed that for the [Sortup](https://github.com/Barsonax/Pathfindax/blob/5af5686de82a6307bb54ffdfd3d4343c32e38d92/Source/Code/Pathfindax/Collections/IndexMaxHeap.cs#L155) method (which plays a crucial role in the perf of the pathfinding) that without the property there are 2 call instructions and with the FCost property there are 4 call instructions (the 2 extra calls here are the calls to the FCost property I believe). It also seems that due to inlining further optimizations are possible which could explain the huge difference in benchmark times as I don't think method call overhead alone is responsible for this.

It seems that the compiler inlines everything except the properties for some reason. Manually making a getter method and adding [MethodImpl(MethodImplOptions.AggressiveInlining)] did not make a difference here. Why is the compiler skipping this optimization when it makes so much of a difference here?

Since iam not that good with assembly code (read: I have zero experience) you can find the generated assembly here:
- [WithoutProperty](https://gist.github.com/Barsonax/882c3e1031d4e866e4393529e15b7971)
- [WithProperty](https://gist.github.com/Barsonax/8056a9a4bdbad747528b53651788b0b2)

**Comment by mikedn** [comment]
As far as I can tell you're using .NET Framework x86. The JIT compiler used by that is not the open sourced RyuJIT but an older, legacy JIT. AFAIR that JIT sometimes does have issue inlining trivial methods. However, it's very unlikely that someone will go and fix that.

It would be useful if you could re-run benchmarks using .NET Framework/Core x64 or .NET Core x86.

**Comment by Barsonax** [comment]
> As far as I can tell you're using .NET Framework x86
But it says ryujit [here](![image](https://user-images.githubusercontent.com/19387223/35615809-906171fc-0673-11e8-98e4-c9eb553dddad.png)
) right? Am I misunderstanding the config in benchmark dotnet? Do I have to change something in the config [here]?(https://github.com/Barsonax/Pathfindax/blob/MoreBenchmarks/Source/Code/Pathfindax.Benchmarks/BenchmarkConfigAttribute.cs)

**Comment by mikedn** [comment]
> But it says ryujit here

Yeah, I saw. I'm not very familiar with the output of Benchmark.DotNet but it also says "32 bit LegacyJIT" at the end of the line.

**Comment by Barsonax** [comment]
> Yeah, I saw. I'm not very familiar with the output of Benchmark.DotNet but it also says "32 bit LegacyJIT" at the end of the line.

Jup noticed that too. Very confusing so I made a issue about that [here](https://github.com/dotnet/BenchmarkDotNet/issues/638).

**Comment by Barsonax** [comment]
Reopening this issue as it also seems to happen when using x64 RyuJIT.
When the AstarNode struct is like [this](https://github.com/Barsonax/Pathfindax/blob/046461dc7be9e852f218e76b5541630a8fab3626/Source/Code/Pathfindax/Nodes/AstarNode.cs)

The perf numbers are like this:
![image](https://user-images.githubusercontent.com/19387223/35629312-957570be-069e-11e8-9567-429452bac758.png)

However when you change the Parent, HCost, GCost and Clearance fields to properties the perf numbers get significantly slower:
![image](https://user-images.githubusercontent.com/19387223/35629126-23b79614-069e-11e8-9926-e3b819141e11.png)

**Comment by mikedn** [comment]
> EDIT2: turning HCost to a property will cause CompareTo to be not inlined anymore. AgressiveInlining does fix it though.

Yes, that's exactly what happens. Unfortunately RyuJIT thinks that `CompareTo` is too large to inline when `HCost` is changed to a property.

I suppose @AndyAyersMS may want to take a look at this. I posted the relevant part of a JIT dump of `SortUp` here: https://gist.github.com/mikedn/875b00655cf77a8fb19aa75da25cc6cd

It's kind of unfortunate that even today the JIT has problems inlining trivial property getters/setters.

**Comment by AndyAyersMS** [comment]
RyuJit makes profitability assessments solely by scanning the IL of the potential inlinee. This analysis isn't transitive -- it does not resolve tokens in the callee's IL or look through calls that the inlinee makes to see if they are also good inline candidates. So the jit has at best a limited and conservative view of the profitability of the inlinee.

Heuristics that look more deeply and more comprehensively are possible but given the throughput constraints the jit faces, these can only be used selectively.  Tiered jitting might allow time for more careful examination.

Along the same lines, the fact that tiering has identified a method as frequently called might allow the jit to simply inline more aggressively and not bother with the more detailed analysis. I started down this path over in #15046 and will come back to it fairly soon.

But those possible improvements are a ways off, especially for desktop CLR. In the meantime, you can always use AggressiveInlining to tell the jit things that it arguably should be able to figure out for itself.

**Comment by AndyAyersMS** [comment]
And if you're curious about attempts to improve the current IL based heuristics, you might find [this writeup](https://github.com/AndyAyersMS/PerformanceExplorer/blob/master/notes/notes-aug-2016.md) interesting.

**Comment by mikedn** [comment]
As a side note: .NET's `Compare` "design" is rather unfortunate performance wise. I haven't looked close enough at your code to tell if you have a choice or not (and there's a pretty good chance you don't due to how pervasive `Compare` is in .NET). But if you do have a choice and want best performance then you could try to use a "Less" predicate instead of a "Compare" predicate, it's quite a bit cheaper.

**Comment by Barsonax** [comment]
> As a side note: .NET's Compare "design" is rather unfortunate performance wise.

Did a quick test as the only place where the CompareTo method is used currently is in the maxheap (purely to get the next 'best' node for pathfinding). Iam not interested if a node has a 'equal' value just need to know if its better or not. In early testing I saw a +20% difference so nice find :). Now its like this:
```
if (FCost == other.FCost)
{
       return HCost < other.HCost;
}
return FCost < other.FCost;
```
